### DISCUSSION TITLE
The database does not match the target database: Table is missing

### DISCUSSION THREAD

**Discussion by Sameer330** [serverpod/serverpod#1886] [Q&A] [answered]
I have a class called `User`:
```
class: User
table: users
fields:
  name: String
  image: String
```
All CRUD operations work fine for this class.

I have another class called `Post` which stores a user's posts:
```
class: Post
table: posts
fields:
  userId: int, relation(parent=users)
  text: String
```
While docker was running:

1. I ran the command: `serverpod create-migration --force` 
2. Then `dart run bin/main.dart --apply-migration`

This was the result:
```
D:\ServerPod-Projects\tbknd\tbknd_server>dart bin/main.dart --apply-migration
Failed to parse command line arguments. Using default values. FormatException: Could not find an option named "apply-migration".
SERVERPOD version: 1.2.3, dart: 3.2.4 (stable) (Thu Dec 21 19:13:53 2023 +0000) on "windows_x64", time: 2024-02-08 05:35:50.558663Z
mode: development, role: monolith, logging: normal, serverId: default
WARNING: The database does not match the target database:
 - Table "posts" is missing.
Hint: Did you forget to apply the migrations (--apply-migrations) or run a repair migration (--apply-repair-migration)?
Insights listening on port 8081
Server default listening on port 8080
Webserver listening on port 8082
CPU and memory usage metrics are not supported on this platform.
```

I tried running the endpoint for posts table but it didn't work. 

How to solve this?

`Reading Users is working - OK 200`

![image](https://github.com/serverpod/serverpod/assets/83571153/16b55463-732c-4c8d-9d87-ab5f2cc95439)

`But adding posts does not work - Bad Request 400`

![image](https://github.com/serverpod/serverpod/assets/83571153/af05b3d7-0f83-40f8-8515-d00fa00f5fea)

**Comment by Isakdl** [comment]
From your log `Failed to parse command line arguments. Using default values. FormatException: Could not find an option named "apply-migration".`

The issue is that there should be an `s` at the end. The correct flag is `--apply-migrations` :)

**Reply by Sameer330** [reply]
I tried running the server with the right flag `--apply-migrations` and this was the resulting log:

```
D:\ServerPod-Projects\tbknd\tbknd_server>serverpod create-migration -t "v1.0.0" 
Invalid tag name. Tag names can only contain lowercase letters, number, and dashes.

D:\ServerPod-Projects\tbknd\tbknd_server>serverpod create-migration -t "v1-0-0" 
⠸ Creating migration... (1.0s)
No changes detected.
✗ Creating migration (1.0s)

D:\ServerPod-Projects\tbknd\tbknd_server>serverpod create-migration -t "v1-0-0" --force
✓ Creating migration (0.5s)
 • Migration created: migrations\20240208104102354-v1-0-0
✓ Done.


D:\ServerPod-Projects\tbknd\tbknd_server>dart run bin/main.dart --apply-migrations "v1-0-0" 
SERVERPOD version: 1.2.3, dart: 3.2.4 (stable) (Thu Dec 21 19:13:53 2023 +0000) on "windows_x64", time: 2024-02-08 10:43:06.329684Z
mode: development, role: monolith, logging: normal, serverId: default
Failed to apply database migrations. Exception: Version 20240206171738894 not found in project.
Insights listening on port 8081
Server default listening on port 8080
Webserver listening on port 8082
CPU and memory usage metrics are not supported on this platform.
```
Also, I deleted some of the previous migrations since I think they are just duplicates. I want to know how to run the latest migration. I tried running the latest migration by assigning a tag but it didn't work.

I also tried calling my `post/createPost` endpoint but it returned status code 400 again :(

![image](https://github.com/serverpod/serverpod/assets/83571153/8fc1d6ab-1ee0-4d79-9cd7-6dc2745c9d53)

**Reply by Isakdl** [reply, marked as answer]
I see, the migration command requires the latest applied migration to be in the folder. The error `Exception: Version 20240206171738894 not found in project.` cannot find the version that was last applied in the database. You have a few options to fix this. If you have the history in git you can bring back this version and create a new migration based on that migration. You can also use the repair migration feature, i.e. `serverpod create-repair-migration` and then `dart bin/main.dart --apply-repair-migration`. This will create a special migration that will bring everything in sync.

Or lastly you can also just wipe the database (if you don't care about the data) and apply the migration on a clean database.

**Reply by Sameer330** [reply]
I think I will go for a new project while taking the current models and endpoint. Anyways, thanks @Isakdl for the assistance. 👍🙌

**Reply by Sameer330** [reply]
@Isakdl 
This time, I created a new project and added the code from my previous project. The user endpoint worked fine but the post endpoint returned 400 status code.

These were the logs:

```
D:\ServerPod-Projects\tbknd2\tbknd2_server>dart bin/main.dart --apply-migrations
SERVERPOD version: 1.2.3, dart: 3.2.4 (stable) (Thu Dec 21 19:13:53 2023 +0000) on "windows_x64", time: 2024-02-08 17:36:57.352221Z
mode: development, role: monolith, logging: normal, serverId: default
Applied database migration:
 - 20240208173507023
Insights listening on port 8081
Server default listening on port 8080
Webserver listening on port 8082
CPU and memory usage metrics are not supported on this platform.
METHOD CALL: user.readAll duration: 11ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 24ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 25ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 16ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 17ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 8ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 11ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 19ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 186ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 11ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.create duration: 6ms numQueries: 0 authenticatedUser: null
METHOD CALL: user.readAll duration: 4ms numQueries: 0 authenticatedUser: null
```

The posts table looks like this:
```
class: Post
table: posts
fields:
  userId: int, relation(parent=users)
  text: String
```

and the posts endpoint looks like this:
```
import 'package:serverpod/serverpod.dart';
import 'package:tbknd2_server/src/generated/protocol.dart';

class PostsEndpoint extends Endpoint {
  Future<Map<String, dynamic>> createPost(Session session, Post post) async {
    Map<String, dynamic> res = {};

    print('Creating post: ${post.text} for user: ${post.userId}');

    var newPost = await Post.db.insertRow(
      session,
      post,
    );

    res['statusCode'] = 200;
    res['post'] = newPost.toJson();

    return res;
  }

  Future<Map<String, dynamic>> readAllPosts(Session session) async {
    Map<String, dynamic> res = {};

    res['statusCode'] = 200;
    res['data'] = await Post.db.find(
      session,
    );

    return res;
  }
}
```

The endpoint `posts/readAllPosts` works as expected but `posts/createPost` returns status code 400. 

Is it because of the foreign key relation? How to solve this?

**Reply by Sameer330** [reply]
@Isakdl 

I found the solution to the `400 Bad Request` error.

Turns out I was calling the `post/createPost` endpoint instead of the **`posts/createPost`**

### Always use the table name; not the class name in the endpoints.

**Reply by kaljitism** [reply]
@Sameer330  What do you mean by "Always use the table name; not the class name in the endpoints." This is my code - 

`Future<void> createUser(Session session, User user) async {
    session.log("user: $user");
    await User.db.insertRow(session, user);
  }`

I am using class Name User to access the db, how do I use the table name here???

**Reply by Sameer330** [reply]
I am referring to the API calls that I was trying out in Postman API Client.

What you are referring to is "Accessing the DB within an API method/function"

Let's say you have a model like this in your `*_server/lib/src/models/` directory:

```
class: User
table: users
fields:
  id: int
  name: String
```

and in your `*_server/lib/src/endpoints/` directory, you have a method/function:

```
Future<void> createUser(Session session, User user) async {
  session.log("user: $user"); 
  await User.db.insertRow(session, user); 
}
```

If you want to call this API in your API Client (E.g. - Postman), You have to call the API like this:
![image](https://github.com/serverpod/serverpod/assets/83571153/7063e0ad-a7f5-440f-90b6-07c13b1fa9d4)

## I hope you understand..

**Reply by kaljitism** [reply]
Thanks, I understand..
